### ISSUE TITLE
Implement BinaryHeap

### ISSUE THREAD

**Issue by jsergiu** [Freeyayo/ifoo#21]
I created this issue to let you know that I'm working on implementing a binary heap just so we avoid duplication. 
PR coming up in the next days.

**Comment by Freeyayo** [comment]
Great!

**Comment by jsergiu** [comment]
PR here: https://github.com/Freeyayo/ifoo/pull/23

I plan to add intialization and compareFunction parameter in a future PR.

**Comment by Freeyayo** [comment]
These two days I've been busy, so I just read your implement code and test code today. 

Beautiful implementation, I mean it, bro! 👍👍👍

`toArray()` keeps the same sequence as the binary heap, nice.

If `_swapNodes(a: number, b: number)` could implement like this: `[a,b]=[b,a]`. But it doesn't matter.

There're two things I don't quite understand:
1. How do I make of **Objects in the heap are stored as shallow copies to prevent side effects**?
2. Every time when we `delete` or `pop` a node, we swap the position between the target node and the last node, then do `this._heap[this._length] = undefined`. Why not splice `this._heap(-1)`? Is there anything I misunderstand? 

Thanks!

**Comment by jsergiu** [comment]
Thanks, this one was a bit tricky but fun to implement. 

Good point about _swapNodes but in this case destructing doesn't work because i'm swapping the objects in the heap at indexes a and b, and not the numbers a and b themselves.

1. When an item is pushed or peeked I use the shallowCopy function that simply calls Object.assign({}, item). The reason for this is to avoid external code messing up with the values in the heap.
Without copying this could happen:
`
o = { foo: 'bar', value: 5}
heap.push(o)
o.value = 6 // Ups! Now the value in the heap is 6 instead of 5 and the order of the nodes might not be correct anymore.
`

2. 👍 I thought splice returns a new array but I was wrong. I will make a PR with the change.

**Comment by Freeyayo** [comment]
Nice explanation and I learned a lot from your code about Binary Heap.
There's one unimportant thing I still want to clarify:
I mentioned before about destructing `[a, b]=[b, a]`, I think I just misled you by code like `a`, `b`. My intention is to convert
```js
private _swapNodes(a: number, b: number): void {
    const aux = this._heap[a];
    this._heap[a] = this._heap[b];
    this._heap[b] = aux;
}
```
to
```js
private _swapNodes(a: number, b: number): void {
    [ this._heap[a],  this._heap[b]] = [ this._heap[b],  this._heap[a]]
}
```
Would this pattern be more concise? But as I said, it doesn't matter.

Example:
![image](https://user-images.githubusercontent.com/29268642/87141968-687c7c00-c2d6-11ea-8a3c-771ee03c0a81.png)
